posts/inspecting-values-with-repl.md

```diff
@@ -0,0 +1,181 @@
+# Inspecting values with Nix REPL
+
+When working with Nix expressions, it might be useful to be able to inspect some values involved. In these cases, the Nix REPL can be pretty useful.
+
+## Some specific REPL forms
+
+When you open up a REPL, you can use `:?` to see the specific REPL forms, e.g.:
+
+```bash
+bash> nix repl
+
+Welcome to Nix version 2.2. Type :? for help.
+
+nix-repl> :?
+The following commands are available:
+
+  <expr>        Evaluate and print expression
+  <x> = <expr>  Bind expression to variable
+  :a <expr>     Add attributes from resulting set to scope
+  :b <expr>     Build derivation
+  :i <expr>     Build derivation, then install result into current profile
+  :l <path>     Load Nix expression and add it to scope
+  # ...
+```
+
+The top three entries will be most often useful for us.
+
+### Evaluate
+
+The first is the same as an expression evaluation:
+
+```bash
+nix-repl> 1
+1
+```
+
+### Bind
+
+The second is the same as a `let`-bind in an expression:
+
+```bash
+nix-repl> a = 1
+
+nix-repl> a
+1
+```
+
+This is equivalent to `let a = 1; in ...`
+
+### Add attributes to scope
+
+The third is the same as a `with` in an expression:
+
+```bash
+nix-repl> :a { hello = "bye"; }
+Added 1 variables.
+
+nix-repl> hello
+"bye"
+```
+
+This is the same as `with { hello = "bye"; }; hello`.
+
+Chances are, you probably won't want to use this one often, as you may end up with a scope full of random identifiers you do not want to use. However, this can be useful at least when you are using builtins:
+
+```bash
+nix-repl> :a builtins
+Added 101 variables.
+
+nix-repl> attrNames { hello = 1; bye = 2; }
+[ "bye" "hello" ]
+```
+
+## Using the REPL with `<nixpkgs>`
+
+Using the REPL with nixpkgs is one of the most common usages of the REPL. From here, you can inspect what packages are available, specific packages' versions, and try launching a Nix Shell with these packages.
+
+```bash
+
+bash> nix repl '<nixpkgs>' # or `nix repl your-pinned-nixpkgs.nix`
+
+Loading '<nixpkgs>'...
+Added 10647 variables.
+```
+
+For the most part, what you will find most useful is being able to tab-complete, e.g.:
+
+```bash
+nix-repl> pk<TAB>
+pk2cmd              pkgconfig           pkgsBuildTarget     pkgsStatic
+pkcs11helper        pkgconfigUpstream   pkgsCross           pkgsTargetTarget
+pkg-config          pkgs                pkgsHostHost        pkgsi686Linux
+pkg-configUpstream  pkgsBuildBuild      pkgsHostTarget      pktgen
+pkgconf             pkgsBuildHost       pkgsMusl
+```
+
+### Building derivations
+
+```bash
+nix-repl> :?
```

Comment:
should this section include `:b` ? You describe it later but might be worth making clear that you intend to discuss it in this section.

Reply:
I have used `:b` more than the other commands but that may just be a difference in what I use the repl for.

Reply:
yup, i rearranged some things so this needs to go here too